### PR TITLE
Support DRF / non-model-based choosers

### PR DESCRIPTION
**Pull request by gasman**
Part of #8480; incorporates #8804.

Makes minor changes to the chooser implementation so that they don't fail outright when a model is not specified, then adds documentation for how to override the appropriate methods for building a chooser powered by a non-model data source.

Ideally I'd have liked to provide a generic base implementation for API-based choosers rather than just documenting the things to override - but given how wildly APIs can differ in their JSON structure (if indeed it's JSON at all) and mechanisms for pagination and filtering, I think that would be best handled by a separate utility class that wraps those differences and presents a standard interface for querying that dataset. And, furthermore, that interface should be a strict subset of Django's QuerySet (so that the 95% of choosers that _are_ using models can use the native queryset interface instead of having to adopt a new dumbed-down one). At that point this becomes a whole other project, one that would probably be a worthwhile addition to the Django ecosystem in its own right, but definitely out of scope for the page editor redevelopment project.

### CODE REVIEW

**Comment by gasman** [comment]
Rebased

**Review by lb- — APPROVED** [review]
@gasman  approved but if possible can you update the docs to be a bit clearer on the following;

* The note about DRF 'defaults' is a bit hard to parse - I was a bit confused on the fact that this was needed. Maybe we can change the example to a public API that already exists? https://github.com/public-apis/public-apis#animals OR just call out a more specifically that the example code does not include the API set up
* I could not work out how to actually use the chooser widget, I know the code is there but I could not get the widget to render.
* The code example also does not call out specifically that you must follow the previous steps on the page and put the viewset registration in the hooks file.

As this is an advanced feature, maybe I am just not reading it right, but I think it might be nice to do another pass on the docs here if you get the time for 4.0.

Finally - not a huge issue - `u-hidden` is in the legacy utils, and new code should use `w-hidden`. But that is not clear at all at the moment so I will raise a new issue for this. https://github.com/wagtail/wagtail/issues/8947
